README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# *KRAS* allele-specific CNA <a href="https://jhrcook.github.io/KrasAlleleCna/index.html"> <img src="man/figures/logo.png" align="right" alt="" width="120" /> </a>

[![License: GPL
v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Travis build
status](https://travis-ci.org/jhrcook/KrasAlleleCna.svg?branch=master)](https://travis-ci.org/jhrcook/KrasAlleleCna)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/jhrcook/KrasAlleleCna?branch=master&svg=true)](https://ci.appveyor.com/project/jhrcook/KrasAlleleCna)
[![Coverage
status](https://codecov.io/gh/jhrcook/KrasAlleleCna/branch/master/graph/badge.svg)](https://codecov.io/github/jhrcook/KrasAlleleCna?branch=master)

You can install this R package from GitHub and run the analyses yourself
(the raw data files are not included, though have been loaded into RData
objects, already).

``` r
devtools::install_github("jhrcook/KrasAlleleCna")
library(KrasAlleleCna)
```

A website to peruse this work is available
[here](https://jhrcook.github.io/KrasAlleleCna/). The Vignettes contain
the analyses.

## Purpose

In [Poulin *et al.*
(2019)](http://cancerdiscovery.aacrjournals.org/content/early/2019/04/05/2159-8290.CD-18-1220),
the authors studied the specific oncogenic properties of *KRAS* A146T
mutations, a hotspot-mutation only found in colorectal adenocarcinoma
(COAD). For this paper, I was asked to find out whether there was a
difference in the copy number of the mutant allele between human COAD
patients with G12D (the most common *KRAS* oncogenic mutation) and A146T
mutations. Not only would a difference suggest that dosage is an
important factor, but could also point to differences in interactions
with the wild-type (WT) allele ([Zhou *et al.*,
2016](https://www.ncbi.nlm.nih.gov/pubmed/27422332), [Lin and
Haigis, 2018](https://www.ncbi.nlm.nih.gov/pubmed/29425486)).

This analysis was divided into 4 parts: A) downloading the data, B)
processing the data, C) calculating the allele copy number, and D)
analyzing the results.

## Conclusions

There was no detectable difference between the copy number of the G12D
or A146T *KRAS* alleles.

The below plot showed the copy number of the mutant and WT allele in
G12D and A146T samples.

<img src="man/figures/README-lateralplot-1.png" width="100%" />

Though there seems to be increased amplification in human tumor samples,
the number of available samples limited the power of the study.

Heatmaps of the genotypes of the samples highlighted that the most
common genotype was 1 mutant : 1 WT
allele.

<img src="man/figures/README-genotypeheatmap-1.png" width="50%" /><img src="man/figures/README-genotypeheatmap-2.png" width="50%" />

-----

**Check out more work from the Haigis Lab at our
[website](https://www.haigislab.org)\!**

Any mistakes or questions? Open an
[Issue](https://github.com/jhrcook/KrasAlleleCna/issues) on GitHub.


NEWS.md
# KrasAlleleCna 0.1.1

Minor fixes for some things I overlookd with the first release. Some documentation was improved. Image output in the vignette "D.01 - Plot Allele-specific Copy Number Results" was fixed, too.


# KrasAlleleCna 0.1.0

Finished first round of work on the package. Everything builds and analysis is complete.


# KrasAlleleCna 0.0.0.9000

Beginning Project-as-a-Package for KRAS CNA analysis.



tests/testthat.R
library(testthat)
library(KrasAlleleCna)

test_check("KrasAlleleCna")


tests/testthat/test-parse_vcf.R
context("test-parse_vcf")

test_that("VCF files are properly parsed", {
    vcf_example <- "005ea64b-18e7-4f50-b940-38322a96e034_mpileup.vcf"
    vcf <- parse_vcf(vcf_example)
    expect_true(all(c("start", "ref", "alt", "num_high_qual_bases", "allelic_depths") %in% colnames(vcf)))
})


tests/testthat/test-utilities.R
context("test-utilities")

test_that("number of unique vals works", {
  expect_equal(n_unique(c(1, 2, 3, 4)), 4)
  expect_equal(n_unique(c(1, 2, 3, 4, 4)), 4)
  expect_equal(n_unique(c(1, 2, 3, 4, NA)), 5)
  expect_equal(n_unique(c("A", "B", "C", "A")), 3)
})


R/data.R
#' TCGA Sample Sheet
#'
#' Contains the information of the files in the manifest. The column names
#'     were changed to a tidy format (lowercase and spaces)
#'
#' @format a tibble (1488 x 5)
#' \describe{
#'     \item{file_id}{file ID}
#'     \item{file_name}{file name}
#'     \item{data_category}{category of data (are all "Sequencing Reads")}
#'     \item{data_type}{type of data (are all "Aligned Reads")}
#'     \item{project_id}{ID of the source project (TCGA-COAD or -READ)}
#'     \item{case_id}{case ID}
#' }
#'
#' @source \url{https://portal.gdc.cancer.gov}
"tidy_tcga_sample_sheet"



#' ANNOVAR results
#'
#' A tibble of the annotation results from ANNOVAR. It was joined with the
#'     sample information. Get more information on the returned values from
#'     the \href{http://annovar.openbioinformatics.org/en/latest/}{ANNOVAR documentation}
#'
#' @format a tibble (411 x 19)
#' \describe{
#'     \item{file_id}{file ID}
#'     \item{Chr}{chromosome}
#'     \item{Start}{start of variant}
#'     \item{End}{end of variant}
#'     \item{Ref}{reference allele}
#'     \item{Alt}{alternate allele}
#'     \item{Func.refGene}{where the alteration is}
#'     \item{Gene.refGene}{gene name}
#'     \item{GeneDetail.refGene}{}
#'     \item{ExonicFunc.refGene}{type of mutation}
#'     \item{AAChange.refGene}{amino acid change}
#'     \item{aa_mod}{more usable amino acid change}
#'     \item{file_name}{name of the file}
#'     \item{data_category}{category of data (are all "Sequencing Reads")}
#'     \item{data_type}{type of data (are all "Aligned Reads")}
#'     \item{project_id}{ID of the source project (TCGA-COAD or -READ)}
#'     \item{case_id}{case ID}
#'     \item{sample_id}{sample ID}
#'     \item{sample_type}{from where the sample was taken}
#' }
#'
"annovar_tib"


#' mpileup and ANNOVAR data merged
#'
#' A tibble of both mpileup and ANNOVAR data - thus it holds the number of
#'     reads at each mutations
#'
#' @format a tibble (190 x 23)
#' \describe{
#'     \item{file_id}{file ID}
#'     \item{Chr}{chromosome}
#'     \item{Start}{start of variant}
#'     \item{End}{end of variant}
#'     \item{Ref}{reference allele}
#'     \item{Alt}{alternate allele}
#'     \item{Func.refGene}{where the alteration is}
#'     \item{Gene.refGene}{gene name}
#'     \item{GeneDetail.refGene}{}
#'     \item{ExonicFunc.refGene}{type of mutation}
#'     \item{AAChange.refGene}{amino acid change}
#'     \item{aa_mod}{more usable amino acid change}
#'     \item{total_num_reads}{total number of reads}
#'     \item{num_mut_reads}{number of mutant allele reads}
#'     \item{file_name}{name of the file}
#'     \item{data_category}{category of data (are all "Sequencing Reads")}
#'     \item{data_type}{type of data (are all "Aligned Reads")}
#'     \item{project_id}{ID of the source project (TCGA-COAD or -READ)}
#'     \item{case_id}{case ID}
#'     \item{sample_id}{sample ID}
#'     \item{sample_type}{from where the sample was taken}
#'     \item{keep}{which bam file to keep}
#'     \item{downloaded}{logical for if the file was downloaded successfully}
#' }
#'
"allele_depth_tib"


#' All data files merged
#'
#' @description Contains ANNOVAR mutation calls with the number of reads from
#'     'mpileup', copy number of the locus, and purity of the tumor sample.
#'     The allele-specific copy number was calculated from those values for
#'     each KRAS allele.
#'
#' @format a tibble (182 x 25)
#' \describe{
#'     \item{common_id}{ID shared by all data tables}
#'     \item{Chr}{chromosome}
#'     \item{Start}{start of variant}
#'     \item{End}{end of variant}
#'     \item{aa_mod}{more usable amino acid change}
#'     \item{total_num_reads}{total number of reads at this site}
#'     \item{num_mut_reads}{number of mutant reads}
#'     \item{Ref}{reference allele}
#'     \item{Alt}{alternate allele}
#'     \item{Func.refGene}{where the alteration is}
#'     \item{Gene.refGene}{gene name}
#'     \item{ExonicFunc.refGene}{type of mutation}
#'     \item{file_name}{name of the file}
#'     \item{file_id}{file ID}
#'     \item{project_id}{ID of the source project (TCGA-COAD or -READ)}
#'     \item{case_id}{case ID}
#'     \item{sample_id}{sample ID}
#'     \item{sample_type}{from where the sample was taken}
#'     \item{CNV_start}{start of the CNV call}
#'     \item{CNV_end}{end of the CNV call}
#'     \item{copy_number}{copy number at the locus}
#'     \item{purity}{purity of the tumor sample}
#'     \item{cn_mut}{copy number of the mutant allele}
#'     \item{copy_number_adj}{copy number of the locus, adjusted for purity}
#'     \item{cn_wt}{copy number of the WT allele}
#' }
"allele_data_filt"


#' CNA information from TCGA
#'
#' @description This tibble contains the CNV information on all samples in
#'     TCGA. It was downloaded from
#'     \url{https://gdc.cancer.gov/about-data/publications/pancanatlas}.
#'
#' @format a tibble (2539963 x 6)
#' \describe{
#'     \item{Sample}{sample ID}
#'     \item{Chromosome}{chromosome of the CNA}
#'     \item{Start}{start of the CNA}
#'     \item{End}{end of the CNA}
#'     \item{Num_Probes}{number of probes over the region}
#'     \item{Segment_Mean}{the CNA value equivalent to log2(copy-number/ 2)}
#' }
"cnv_tib"


R/tidy_annovar.R
#' Parse ANNOVAR mutation
#'
#' Reads and parses an ANNOVAR file to return the amino acid change in a more
#'     understandable way in the column named \code{aa_mod}.
#'
#' @param anno_tib tibble of the ANNOVAR file
#'
#' @return Returns the tibble, checking if it is empty.
#'     If the file was empty, an empty tibble is returned. A column is added
#'     that has a simple version of the amino acid change \code{aa_mod}.
#'
#' @examples
#' extdata_dir <- system.file("extdata", package = "KrasAlleleCna")
#' annovar_file <- readRDS(file.path(extdata_dir, "annovar_output_list.rds"))[[3]]
#' annovar_file
#'
#' parsed_annovar <- parse_annovar_mutation(annovar_file)
#' parsed_annovar
#'
#' # the new column has easy to use amino acid mutations
#' parsed_annovar$aa_mod
#'
#' 
#' @importFrom magrittr %>% %<>%
#' @export parse_annovar_mutation
parse_annovar_mutation <- function(anno_tib) {
    if (nrow(anno_tib) == 0) {
        # if empty, set to all NAs and continue in pipeline
        anno_tib <- tibble::tibble(Chr = NA,
                                   Start = NA,
                                   End = NA,
                                   Ref = NA,
                                   Alt = NA,
                                   Func.refGene = NA,
                                   Gene.refGene = NA,
                                   GeneDetail.refGene = NA,
                                   ExonicFunc.refGene = NA,
                                   AAChange.refGene = NA) %>%
            dplyr::slice(0)
    }
    # parse amino acid modification (aa_mod)
    anno_tib %<>%
        dplyr::mutate(aa_mod = stringr::str_extract(AAChange.refGene,
                                                    "(?<=p\\.)[:alnum:]+"),
               aa_mod = ifelse(is.na(Chr), "WT", aa_mod))
    return(anno_tib)
}

# for "parse_annovar_mutation"
utils::globalVariables(c("AAChange.refGene", "Chr", "aa_mod"), add = TRUE)


R/download-data.R
#' Select the best filename to keep
#'
#' @description Some cases in the sample sheet have multiple files.
#'     `choose_one_filename()`uses a set of (fairly arbitrary) checks to select
#'     one filename. `two_filename_handler()` and `three_filename_handler()`
#'     parse two or three file names based on my heuristics for which is most
#'     likely to be more useful.
#'
#' @param fn vector of filenames
#'
#' @return a single filename from the input vector that should be retained
#'
#' @examples
#' test_fns <- c("TCGA-A6-2672-01B-03D-2301-10_Illumina_gdc_realn.bam",
#'               "C982.TCGA-A6-2672-01B-03D-2298-08.2_gdc_realn.bam")
#' choose_one_filename(test_fns)
#'
#' @importFrom stringr str_detect str_subset
#' @importFrom magrittr %>%
#' @export choose_one_filename
choose_one_filename <- function(fn) {
    # return singles
    if (length(fn) == 1) {
        return(fn)
    }

    if (length(fn) == 2) {
        keep_fn <- two_filename_handler(fn)
    } else if (length(fn) == 3) {
        keep_fn <- three_filename_handler(fn)
    } else {
        keep_fn <- fn
    }

    if (length(keep_fn) == 1) {
        return(keep_fn)
    } else {
        cat_multiple_file_msg(keep_fn)
        return(keep_fn[[1]])
    }

}

#' @rdname choose_one_filename
#' @importFrom stringr str_detect str_subset
#' @export three_filename_handler
two_filename_handler <- function(fn) {
    if (sum(str_detect(fn, "^TCGA")) == 1) {
        keep_fn <- str_subset(fn, "^TCGA")
    } else if (sum(str_detect(fn, "-[:digit:]{2}D-")) == 1 &
               sum(str_detect(fn, "-[:digit:]{2}W-")) == 1) {
        keep_fn <- str_subset(fn, "-[:digit:]{2}D-")
    } else if (sum(str_detect(fn, "hg19_Illumina_gdc_realn.bam$")) == 1) {
        keep_fn <- fn[!str_detect(fn, "hg19_Illumina_gdc_realn.bam$")]
    } else if (any(str_detect(fn, "-01A-"))) {
        keep_fn <- str_subset(fn, "-01A-")
    } else {
        keep_fn <- fn
    }
    return(keep_fn)
}

#' @rdname choose_one_filename
#' @importFrom stringr str_detect str_subset
#' @export three_filename_handler
three_filename_handler <- function(fn) {
    if (sum(str_detect(fn, "hg19_Illumina_gdc_realn.bam$")) == 1) {
        keep_fn <- fn[!str_detect(fn, "hg19_Illumina_gdc_realn.bam$")]
    } else {
        keep_fn <- fn
    }

    if (length(keep_fn) == 2) {
        keep_fn <- two_filename_handler(keep_fn)
    }
    return(keep_fn)
}

# print a message that there is more than one file and list file names
cat_multiple_file_msg <- function(fn) {
    cat("multiple files:\n\t", paste(fn, collapse = "\n\t"), "\n", sep = "")
}


R/heatmap_genotype.R
#' Plot genotype heatmaps
#'
#' @description Plot the genotypes of mutant samples by allele as a heatmap.
#'     See vignette "D.01 - Plot Allele-specific Copy Number Results" for how
#'     to use this function. It was made into a function primarily so it could
#'     also be used in the README.
#'
#' @param tib copy number information
#' @param allele allele to plot for
#'
#' @import dplyr
#' @import ggplot2
#' @importFrom rlang !!
#' @export plot_genotype_heatmap


plot_genotype_heatmap <- function(tib, allele) {
    hm <- tib %>%
        filter(aa_mod == !!allele) %>%
        mutate(cn_mut_round = round(cn_mut),
               cn_wt_round = round(cn_wt)) %>%
        group_by(aa_mod, cn_mut_round, cn_wt_round) %>%
        summarise(num = n_distinct(common_id)) %>%
        ungroup() %>%
        group_by(aa_mod) %>%
        mutate(freq = num / n()) %>%
        ungroup() %>%
        stats::na.omit() %>%
        ggplot(aes(x = cn_wt_round, y = cn_mut_round)) +
        geom_tile(aes(fill = freq)) +
        geom_label(aes(label = num),
                   fill = "grey35", color = "white",
                   fontface = "bold", size = 5) +
        scale_fill_gradient(low = "blue", high = "red") +
        theme(panel.grid.major = element_blank(),
              panel.grid.minor = element_line(color = "black",
                                              size = 0.5),
              legend.position = "none",
              panel.background = element_rect(fill = "white"),
              panel.border = element_rect(color = "black",
                                          fill = NA,
                                          size = 0.5),
              axis.ticks = element_blank()) +
        labs(title = paste("Genotype heatmap for", allele, "mutants"),
             x = "WT allele copy number",
             y = "mutant allele copy number") +
        scale_y_continuous(expand = c(0, 0),
                           breaks = c(0, 1, 2, 3),
                           minor_breaks = seq(0.5, 3.5, 0.5)) +
        scale_x_continuous(expand = c(0, 0),
                           breaks = c(0, 1, 2, 3),
                           minor_breaks = seq(0.5, 3.5, 0.5))
    return(hm)
}

utils::globalVariables(
    c("cn_mut",
      "cn_mut_round",
      "cn_wt",
      "cn_wt_round",
      "common_id",
      "freq",
      "num"),
    add = TRUE)


bash-scripts/download_process_tcga_bams.sh
#!/bin/bash
#!/bin/bash
#SBATCH -p short
#SBATCH -t 0-0:03
#SBATCH --mem 1G
#SBATCH -c 1
#SBATCH --mail-type ALL
#SBATCH -e array_job_logs/slurm-%A_%a.log
#SBATCH -o array_job_logs/slurm-%A_%a.log

module load gcc samtools/1.9 bcftools/1.9 annovar/20170601

# constants
# reference genome from TCGA
TCGA_REF_38=/n/no_backup2/dbmi/park/jc604/databases/reference_genomes/GRCh38.d1.vd1.fa
# read in the GDC token to access the data
GDC_TOKEN=$(<~/gdc_tokens/gdc-user-token.2019-03-04.txt)
# location of ANNOVAR databases
ANNOVAR_DB=/n/no_backup2/dbmi/park/jc604/databases/annovar_db

# get file name to download from GDC
	# $SLURM_ARRAY_TASK_ID --> array number, used as line number here
	# $1 --> file to get GDC file name from
GDC_FILE=$(sed -n "${SLURM_ARRAY_TASK_ID}p" $1)

# other file names
BAM=tmp-files/${GDC_FILE}.bam
CALL_VCF=tmp-files/${GDC_FILE}_call.vcf
MPILEUP_VCF=data-raw/gdc/mpileup_output/${GDC_FILE}_mpileup.vcf
AVINPUT=tmp-files/${GDC_FILE}.avi
AVOUTPUT=data-raw/gdc/annovar_output/${GDC_FILE}

# get BAM slice of KRAS from GDC
curl \
    --header "X-Auth-Token: $GDC_TOKEN" "https://api.gdc.cancer.gov/slicing/view/${GDC_FILE}?region=chr12:25200000-25250000" \
    --output $BAM

# check if BAM is the correct format
# else print GDC_FILE to std err
if [[ "$(file $BAM)" != "${BAM}: gzip compressed data, extra field" ]]; then
	echo "${GDC_FILE}: downloaded file not a BAM" >> tmp-files/failed_downloads.txt
	exit
fi

# make BAM index file
samtools index $BAM

# call mutations
bcftools mpileup \
	-Ou \
	--fasta-ref $TCGA_REF_38 \
	$BAM | \
bcftools call \
	-m \
	-Ov \
	--output $CALL_VCF

# make annotated mpileup VCF file
bcftools mpileup \
	-Ov \
	--annotate 'INFO/AD,FORMAT/AD,FORMAT/DP' \
	--fasta-ref $TCGA_REF_38 \
	--output $MPILEUP_VCF \
	$BAM

# annotate mutations
convert2annovar.pl \
    --format vcf4 \
    --outfile $AVINPUT \
    $CALL_VCF
table_annovar.pl \
    -buildver hg38 \
    --outfile $AVOUTPUT \
    -remove \
    -protocol refGene \
    -operation g \
    -nastring NA \
    $AVINPUT \
    $ANNOVAR_DB

# remove intermediate files
rm $BAM $CALL_VCF $AVINPUT ${BAM}.bai


# Run this script as a batch array with the following command:
# sbatch --array=1-$(wc -l < input/tcga_filename_list.txt)%50 download_process_tcga_bams.sh input/tcga_filename_list.txt


R/parse_vcf.R
#' Parse a VCF file
#'
#' @description Parse the read-depth information of a VCF file produced by
#'     mpileup
#'
#' @param fname path to the file
#'
#' @return a tibble
#'
#' @examples
#' \dontrun{
#' # head of the VCF file
#' #> ##fileformat=VCFv4.2
#' #> ##FILTER=<ID=PASS,Description="All filters passed">
#' #> ##bcftoolsVersion=1.9+htslib-1.9
#' #> ##bcftoolsCommand=mpileup -Ov --annotate INFO/AD,FORMAT/AD,FORMAT/DP ...
#' #> ##reference=file:///n/no_backup2/dbmi/park/jc604/databases/reference_...
#' #> ##contig=<ID=chr1,length=248956422>
#' #> ##contig=<ID=chr2,length=242193529>
#' #> ##contig=<ID=chr3,length=198295559>
#' #> ##contig=<ID=chr4,length=190214555>
#' #> ##contig=<ID=chr5,length=181538259>
#' #> ##contig=<ID=chr6,length=170805979>
#' #> ##contig=<ID=chr7,length=159345973>
#' #> ##contig=<ID=chr8,length=145138636>
#' #> ##contig=<ID=chr9,length=138394717>
#' #> ##contig=<ID=chr10,length=133797422>
#' #> ##contig=<ID=chr11,length=135086622>
#' #> ##contig=<ID=chr12,length=133275309>
#' #> ##contig=<ID=chr13,length=114364328>
#' #> ##contig=<ID=chr14,length=107043718>
#' #> ##contig=<ID=chr15,length=101991189>
#' #> ##contig=<ID=chr16,length=90338345>
#' #> ##contig=<ID=chr17,length=83257441>
#' #> ##contig=<ID=chr18,length=80373285>
#' #> ##contig=<ID=chr19,length=58617616>
#' #> ##contig=<ID=chr20,length=64444167>
#' #> ##contig=<ID=chr21,length=46709983>
#' #> ##contig=<ID=chr22,length=50818468>
#' #> ##contig=<ID=chrX,length=156040895>
#' #> ##contig=<ID=chrY,length=57227415>
#' #> ##contig=<ID=chrM,length=16569>
#' #> ##contig=<ID=chr1_KI270706v1_random,length=175055>
#' #> ##contig=<ID=chr1_KI270707v1_random,length=32032>
#' #> ##contig=<ID=chr1_KI270708v1_random,length=127682>
#' #> ...
#' #> ##INFO=<ID=QS,Number=R,Type=Float,Description="Auxiliary tag used for calling">
#' #> ##FORMAT=<ID=PL,Number=G,Type=Integer,Description="List of Phred-scaled genotype likelihoods">
#' #> ##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Number of high-quality bases">
#' #> ##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths">
#' #> ##INFO=<ID=AD,Number=R,Type=Integer,Description="Total allelic depths">
#' #> #CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  TCGA-FZ-5924-01A-13D-1609-08
#' #> chr12   25202434        .       A       <*>     0       .       DP=1;AD=1,0;I16=1,0,0,0,29,841,0,0,60,3600,0,0,0,0,0,0;QS=1,0;MQ0F=0    PL:DP:AD        0,3,29:1:1,0
#' #> chr12   25202435        .       G       <*>     0       .       DP=1;AD=1,0;I16=1,0,0,0,31,961,0,0,60,3600,0,0,1,1,0,0;QS=1,0;MQ0F=0    PL:DP:AD        0,3,31:1:1,0
#' #> chr12   25202436        .       C       <*>     0       .       DP=1;AD=1,0;I16=1,0,0,0,28,784,0,0,60,3600,0,0,2,4,0,0;QS=1,0;MQ0F=0    PL:DP:AD        0,3,28:1:1,0
#' #> chr12   25202437        .       A       <*>     0       .       DP=1;AD=1,0;I16=1,0,0,0,30,900,0,0,60,3600,0,0,3,9,0,0;QS=1,0;MQ0F=0    PL:DP:AD        0,3,30:1:1,0
#' #> chr12   25202438        .       G       <*>     0       .       DP=1;AD=1,0;I16=1,0,0,0,34,1156,0,0,60,3600,0,0,4,16,0,0;QS=1,0;MQ0F=0  PL:DP:AD        0,3,34:1:1,0
#' #> chr12   25202439        .       G       <*>     0       .       DP=1;AD=1,0;I16=1,0,0,0,33,1089,0,0,60,3600,0,0,5,25,0,0;QS=1,0;MQ0F=0  PL:DP:AD        0,3,33:1:1,0
#' #> ...
#'
#' parse_vcf("example.vcf")
#' #> # A tibble: 8,917 x 7
#' #>       start ref   alt    qual read_info    num_high_qual_bas… allelic_depths
#' #>       <dbl> <chr> <chr> <dbl> <chr>        <chr>              <chr>
#' #>  1 25202434 A     <*>       0 0,3,29:1:1,0 1                  1
#' #>  2 25202435 G     <*>       0 0,3,31:1:1,0 1                  1
#' #>  3 25202436 C     <*>       0 0,3,28:1:1,0 1                  1
#' #>  4 25202437 A     <*>       0 0,3,30:1:1,0 1                  1
#' #>  5 25202438 G     <*>       0 0,3,34:1:1,0 1                  1
#' #>  6 25202439 G     <*>       0 0,3,33:1:1,0 1                  1
#' #>  7 25202440 G     <*>       0 0,3,34:1:1,0 1                  1
#' #>  8 25202441 G     <*>       0 0,3,33:1:1,0 1                  1
#' #>  9 25202442 A     <*>       0 0,3,32:1:1,0 1                  1
#' #> 10 25202443 T     <*>       0 0,3,31:1:1,0 1                  1
#' #> # … with 8,907 more rows
#'
#' }
#'
#' @importFrom stringr str_split_fixed str_remove_all
#' @export parse_vcf
parse_vcf <- function(fname) {
    # find the start of the data in the VCF file and read
    header_line <- grep("#CHROM", readLines(fname)) - 1
    v <- readr::read_tsv(fname,
                         col_type = "cdcccdcccc",
                         skip = header_line)
    colnames(v) <- c("chr", "start", "id", "ref", "alt", "qual", "filter",
                     "info", "format", "read_info")
    # breakout the information held in FORMAT
    # FORMAT = PL:DP:AD
        #  x PL = "List of Phred-scaled genotype likelihoods"
        #  * DP = "Number of high-quality bases"
        #  * AD = "Allelic depths"
    v %<>%
        dplyr::select(start, ref, alt, qual, read_info) %>%
        dplyr::mutate(num_high_qual_bases = str_split_fixed(read_info, ":", 3)[, 2],
                      allelic_depths = str_split_fixed(read_info, ":", 3)[, 3],
                      allelic_depths = str_remove_all(allelic_depths, ",0$"))
    return(v)
}



utils::globalVariables(
    c("start",
      "ref",
      "alt",
      "qual",
      "read_info",
      "allelic_depths"),
    add = TRUE
)
